src/promptflow-tools/promptflow/tools/common.py

```diff
@@ -575,85 +618,131 @@ def to_bool(value) -> bool:
+        print(f"Exception occurs: {type(e).__name__}: {unescaped_exception_message}", file=sys.stderr)
+        error_message = f"Failed to render jinja template: {type(e).__name__}: {unescaped_exception_message}. " \
                         + "Please modify your prompt to fix the issue."
         raise JinjaTemplateError(message=error_message) from e
 
 
-def build_escape_dict(kwargs: dict):
-    escape_dict = {}
-    for _, value in kwargs.items():
-        escape_dict = _build_escape_dict(value, escape_dict)
-    return escape_dict
-
+class Escaper:
+    @staticmethod
+    def merge_escape_mapping_of_prompt_results(**kwargs):
+        escape_dict = {}
+        for _, v in kwargs.items():
+            if isinstance(v, PromptResult) and v.need_to_escape():
+                escape_dict.update(v.get_escape_mapping())
+        return escape_dict
+
+    @staticmethod
+    def build_flow_inputs_escape_dict(inputs_to_escape: list, **kwargs):
+        escape_dict = {}
+        if not inputs_to_escape:
+            return escape_dict
+
+        for k, v in kwargs.items():
+            if k in inputs_to_escape:
+                escape_dict = Escaper._build_flow_input_escape_dict(v, escape_dict)
+        return escape_dict
+
+    @staticmethod
+    def _build_flow_input_escape_dict(val, escape_dict: dict):
+        """
+        Build escape dictionary with roles as keys and uuids as values.
+        """
+        if isinstance(val, ChatInputList):
+            for item in val:
+                Escaper._build_flow_input_escape_dict(item, escape_dict)
+        elif isinstance(val, str):
+            pattern = r"(?i)^\s*#?\s*(" + "|".join(VALID_ROLES) + r")\s*:\s*\n"
+            roles = re.findall(pattern, val, flags=re.MULTILINE)
+            for role in roles:
+                if role not in escape_dict:
+                    # We cannot use a hard-coded hash str for each role, as the same role might be in various case.
+                    # For example, the 'system' role may vary in input as 'system', 'System', 'SysteM','SYSTEM', etc.
+                    # To convert the escaped roles back to original str, we need to use different uuids for each case.
+                    escape_dict[role] = str(uuid.uuid4())
+
+        return escape_dict
+
+    @staticmethod
+    def _escape_roles_in_flow_input(val, escape_dict: dict):
+        """
+        Escape the roles in the prompt inputs to avoid the input string with pattern '# role' get parsed.
+        """
+        if not escape_dict:
+            return val
+
+        if isinstance(val, ChatInputList):
+            return ChatInputList([Escaper._escape_roles_in_flow_input(item, escape_dict) for item in val])
+        elif isinstance(val, str):
+            for role, encoded_role in escape_dict.items():
+                val = val.replace(role, encoded_role)
+            return val
+        else:
+            return val
+
+    @staticmethod
+    def _unescape_roles(val, escape_dict: dict):
```

Comment:
What if it is a complicated dict with several levels?

Reply:
So far it can not happen. There are only two cases: 1. the val is a string, which is a error message. We need to unescape roles before the error message get presented to user. 2. the val is from a parsed message array, Then the val would either be a string or a list